### PR TITLE
Improvements to swagger to Ballerina code generator

### PR DESCRIPTION
**Pull request by praminda**
## Goals
- Introduce connector generation from API definition
- Introduce struct generation for an API definition
- Update swagger annotations for oas3
- Introduce support for dir as output parameter
- Update skeleton generation templates
- Introduce oas3 security requirement support to codegen
- Introduce support for resource level documentation
- Use deprecated {} and `discontinued` attribute

## Automation tests
 - Unit tests 
   > Yes

## Samples
`ballerina swagger skeleton /home/foo/petstore.yaml -o /home/foo/` will generate a ballerina service skeleton in /home/foo/

### CODE REVIEW

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java [hunk 1]
```diff
@@ -35,41 +39,100 @@
...
+     *
+     * @param openAPI <code>OpenAPI</code> definition object with schema definition
+     */
+    private void setSchemas(OpenAPI openAPI) {
+        this.schemas = new LinkedHashSet<>();
+        Map<String, Schema> schemaMap;
+        if (openAPI.getComponents() == null) {
+            return;
+        }
+        if (openAPI.getComponents().getSchemas() == null) {
```
Comment (nadeeshaan) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: Can we merge these two if conditions together?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java [hunk 2]
```diff
@@ -82,15 +145,32 @@ public OpenApiWrapper buildFromOpenAPI(OpenAPI openAPI) throws MalformedURLExcep
...
+    }
+
+    /**
+     * Extract security requirements as a set.
+     *
+     * @param openAPI <code>OpenAPI</code> definition object with security definition
+     */
+    private void setSecurityRequirements(OpenAPI openAPI) {
+        this.security = new LinkedHashSet<>();
+        List<SecurityRequirement> requirements = openAPI.getSecurity();
```
Comment (sanjeewa-malalgoda) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: Doing null check and checking requirements has element in same line and process would be more human readable. If that failed then we can return.

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import io.swagger.oas.models.media.Schema;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This class wraps the {@link Schema} from swagger models inorder to overcome complications
+ * while populating handlebar templates.
+ */
+public class BalSchema {
```
Comment (sanjeewa-malalgoda) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java]: BallerinaSchema?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java [hunk 4]
```diff
@@ -35,41 +39,100 @@
...
+     * Extract host information from OpenAPI server list.
+     *
+     * @param openAPI <code>OpenAPI</code> definition object with host information
+     */
+    private void setHostInfo(OpenAPI openAPI) throws MalformedURLException {
+        List<Server> serverList = openAPI.getServers();
+        if (serverList == null) {
+            this.host = "localhost";
+            this.port = 80;
+            this.httpsPort = 443;
```
Comment (Manuri) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: Can we use constants for these port values?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java [hunk 5]
```diff
@@ -35,41 +39,100 @@
  * Wrapper for <code>io.swagger.oas.models.OpenAPI</code>
  * <p>This class can be used to push additional context variables for handlebars</p>
  */
-public class OpenApiWrapper {
-
+public class BalOpenApi {
```
Comment (sanjeewa-malalgoda) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: I feel OpenAPIWrapper is more suitable for this. Any thoughts?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java (hunk 2 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: @sanjeewa-malalgoda Did you mean to check if requirements list is not empty?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java (hunk 5 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: This model class is moving bit away from just being a wrapper for OpenAPI type. Shall we use **BallerinaOpenApi**?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+            // change conflicting swagger data types to ballerina data types
+            for (Map.Entry entry : entries) {
+                Schema prop = (Schema) entry.getValue();
+
+                // Extract reference type objects
+                if (prop.get$ref() != null) {
+                    String type = prop.get$ref().substring(prop.get$ref().lastIndexOf("/") + 1);
+                    prop.setType(type);
+                } else {
```
Comment (lankavitharana) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java]: shall we remove this else as well?(continue if)

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java [hunk 7]
```diff
@@ -82,15 +145,32 @@ public OpenApiWrapper buildFromOpenAPI(OpenAPI openAPI) throws MalformedURLExcep
...
+     */
+    private void setSecurityRequirements(OpenAPI openAPI) {
+        this.security = new LinkedHashSet<>();
+        List<SecurityRequirement> requirements = openAPI.getSecurity();
+        if (requirements == null) {
+            return;
+        }
 
-        return this;
+        for (SecurityRequirement requirement: requirements) {
```
Comment (lankavitharana) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: can we use requirements.foreach here?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java (hunk 5 above)
Comment (sanjeewa-malalgoda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: +1

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java (hunk 1 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: Fixed with 06d87e1

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/exception/BalOpenApiException.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ballerinalang.swagger.exception;
+
+/**
+ * Exception type definition for OpenAPI to BalOpenAPI parsing related errors.
+ */
+public class BalOpenApiException extends Exception {
```
[the remark this answers is not in the document]
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/exception/BalOpenApiException.java]: Fixed with 06d87e1
Renamed to `BallerinaOpenApiException`

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java (hunk 2 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: Fixed with 06d87e1

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java (hunk 3 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java]: Fixed with 06d87e1

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java (hunk 4 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: Fixed with 06d87e1

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    private boolean isArraySchema = false;
+
+    public BalSchema buildFromSchema(Schema schema, Map<String, Schema> allSchemas) {
+        this.oasSchema = schema;
+
+        // Keep track of array type schema definitions
+        if (schema instanceof ArraySchema) {
+            this.isArraySchema = true;
+            this.properties = null;
+        } else {
```
[the remark this answers is not in the document]
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java]: Fixed with 06d87e1

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java (hunk 6 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalSchema.java]: Fixed with 06d87e1

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java (hunk 7 above)
Comment (praminda) [reply misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BalOpenApi.java]: Fixed with 06d87e1

**Comment by praminda** [comment]
Updated service generation to match with new syntax. Connector generation is disabled temporarily until core implementation is finalized.

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BallerinaServer.java [hunk 10]
```diff
@@ -0,0 +1,116 @@
+package org.ballerinalang.swagger.model;
```
Comment (djkevincr) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BallerinaServer.java]: Shall we add license header?

File: misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BallerinaSwaggerObject.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package org.ballerinalang.swagger.model;
```
Comment (djkevincr) [inline misc/swagger-ballerina/modules/swagger-to-ballerina-generator/src/main/java/org/ballerinalang/swagger/model/BallerinaSwaggerObject.java]: Shall we add License header?

**Pull request merged**